File: examples/python/robust_curve_fitting.py
"""
Robust curve fitting example

In this example we fit a curve defined by
    y = exp(m * x + c)
The error function is
    \sum_i f(|y_i - exp(m * x_i + c)|^2)
The loss function f can be either identity or robust Cauchy loss function
"""

from __future__ import print_function

from minisam import *
import numpy as np
import math

import matplotlib.pyplot as plt


# ########################## plot and loading utils ############################

# load x-y value pairs
def loadFromFile(filename):
    data = []
    f = open(filename, "r")
    for line in f:
        arr = [float(num) for num in line.split(' ')]
        data.append(np.array(arr))
    return data

# plot raw x-y points
def plotPoints(data, color, r=10):
    xs, ys = [], []
    for d in data:
        xs.append(d[0])
        ys.append(d[1])
    plt.scatter(xs, ys, s=r, facecolors='none', edgecolors=color, marker='o')

# plot fitted exp curve
# y = exp(m * x + c);
def plotExpCurve(m, c, x_start, x_end, color, linewidth=1):
    x_step = 0.05
    xs, ys = [], []
    for x in np.arange(x_start, x_end, x_step):
        xs.append(x)
        ys.append(math.exp(m * x + c))
    plt.plot(xs, ys, color, linewidth)


# ################################## factor ####################################

# exp curve fitting factor
class ExpCurveFittingFactor(Factor):
    # ctor
    def __init__(self, key, point, loss):
        Factor.__init__(self, 1, [key], loss)
        self.p_ = point

    # make a deep copy
    def copy(self):
        return ExpCurveFittingFactor(self.keys()[0], self.p_, self.lossFunction())

    # error = y - exp(m * x + c);
    def error(self, variables):
        params = variables.at(self.keys()[0])
        return np.array([self.p_[1] - math.exp(params[0] * self.p_[0] + params[1])])

    # jacobians
    def jacobians(self, variables):
        params = variables.at(self.keys()[0])
        J_e_mc = np.array([[-self.p_[0] * math.exp(params[0] * self.p_[0] + params[1]), 
            -math.exp(params[0] * self.p_[0] + params[1])]])
        return [J_e_mc]


# ################################# example ####################################

# load data
data = loadFromFile("../data/exp_curve_fitting_data.txt")

# use robust (Cauchy) loss function or not
useRobustLoss = True
if useRobustLoss:
    loss = CauchyLoss.Cauchy(1.0)
else:
    loss = None

# build graph
graph = FactorGraph()
for d in data:
    graph.add(ExpCurveFittingFactor(key('p', 0), d, loss))

# init estimation of curve parameters
init_values = Variables()
init_values.add(key('p', 0), np.array([0, 0]))

print("initial curve parameters :", init_values.at(key('p', 0)))

# Use LM method optimizes the initial values
opt_param = LevenbergMarquardtOptimizerParams()
opt_param.verbosity_level = NonlinearOptimizerVerbosityLevel.ITERATION
opt = LevenbergMarquardtOptimizer(opt_param)

values = Variables()
status = opt.optimize(graph, init_values, values)

if status != NonlinearOptimizationStatus.SUCCESS:
    print("optimization error: ", status)

print("opitmized curve parameters :", values.at(key('p', 0)))

# plot
fig, ax = plt.subplots()

plotPoints(data, 'b')
plotExpCurve(0.3, 0.1, data[0][0], data[-1][0], 'k')
params_opt = values.at(key('p', 0))
plotExpCurve(params_opt[0], params_opt[1], data[0][0], data[-1][0], 'r')

ax.set_title('red is optimized and black is ground truth')
plt.axis('equal')
plt.show()

File: examples/cpp/robust_curve_fitting.cpp
/**
 * Robust curve fitting example
 *
 * In this example we fit a curve defined by
 *     y = exp(m * x + c)
 * The error function is
 *     \sum_i f(|y_i - exp(m * x_i + c)|^2)
 * The loss function f can be either identity or robust Cauchy loss function
 */

#include <minisam/core/Eigen.h>  // include when use Eigen vector in optimization
#include <minisam/core/Factor.h>
#include <minisam/core/FactorGraph.h>
#include <minisam/core/LossFunction.h>
#include <minisam/core/Variables.h>
#include <minisam/nonlinear/LevenbergMarquardtOptimizer.h>

#include <fstream>
#include <iostream>

using namespace std;
using namespace minisam;

/* ************************** file loading utils **************************** */

// load x-y value pairs from file
std::vector<Eigen::Vector2d> loadFromFile(const std::string& filename) {
  ifstream file(filename.c_str(), ifstream::in);
  if (!file) {
    throw invalid_argument("ERROR: cannot load file " + filename);
  }
  std::vector<Eigen::Vector2d> points;
  string line;
  while (getline(file, line)) {
    stringstream ss(line);
    double x, y;
    ss >> x >> y;
    points.push_back(Eigen::Vector2d(x, y));
  }
  return points;
}

/* ******************************** factor ********************************** */

// exp curve fitting factor
class ExpCurveFittingFactor : public Factor {
 private:
  Eigen::Vector2d p_;

 public:
  EIGEN_MAKE_ALIGNED_OPERATOR_NEW

  ExpCurveFittingFactor(Key key, const Eigen::Vector2d& point,
                        const std::shared_ptr<LossFunction>& lossfunc)
      : Factor(1, std::vector<Key>{key}, lossfunc), p_(point) {}

  virtual ~ExpCurveFittingFactor() = default;

  // deep copy function
  std::shared_ptr<Factor> copy() const override {
    return std::shared_ptr<Factor>(new ExpCurveFittingFactor(*this));
  }

  // error function
  // clang-format off
  Eigen::VectorXd error(const Variables& values) const override {
    const Eigen::Vector2d& params = values.at<Eigen::Vector2d>(keys()[0]);
    return (Eigen::VectorXd(1) << p_(1) - std::exp(params(0) * p_(0) + params(1)))
        .finished();
  }

  // jacobians function
  std::vector<Eigen::MatrixXd> jacobians(
      const Variables& values) const override {
    const Eigen::Vector2d& params = values.at<Eigen::Vector2d>(keys()[0]);
    return std::vector<Eigen::MatrixXd>{
        (Eigen::MatrixXd(1, 2) <<
            -p_(0) * std::exp(params(0) * p_(0) + params(1)),
            -std::exp(params(0) * p_(0) + params(1)))
        .finished()};
  }
  // clang-format on
};

/* ******************************* example ********************************** */

int main(int argc, char* argv[]) {
  // load data
  string filename = "../../../examples/data/exp_curve_fitting_data.txt";
  if (argc == 2) {
    filename = string(argv[1]);
  } else if (argc > 2) {
    cout << "usage: ./robust_curve_fitting [data_filename.txt]" << endl;
    return 0;
  }
  const auto data = loadFromFile(filename);

  // loss function
  const bool useRobustLossFuction = true;
  std::shared_ptr<LossFunction> loss;
  if (useRobustLossFuction) {
    loss = CauchyLoss::Cauchy(1.0);
  } else {
    loss = nullptr;
  }

  // build graph
  FactorGraph graph;
  for (const auto& d : data) {
    graph.add(ExpCurveFittingFactor(key('p', 0), d, loss));
  }

  // init estimation of curve parameters
  Variables init_values;
  init_values.add(key('p', 0), Eigen::Vector2d(0, 0));
  cout << "initial curve parameters :" << endl
       << init_values.at<Eigen::Vector2d>(key('p', 0)) << endl;

  // optimize!
  LevenbergMarquardtOptimizerParams opt_param;
  opt_param.verbosity_level = NonlinearOptimizerVerbosityLevel::ITERATION;
  LevenbergMarquardtOptimizer opt(opt_param);

  Variables values;
  opt.optimize(graph, init_values, values);
  cout << "opitmized curve parameters :" << endl
       << values.at<Eigen::Vector2d>(key('p', 0)) << endl;

  return 0;
}
